pallets/domains/src/lib.rs

```diff
@@ -411,58 +362,81 @@ pub mod pallet {
 
             RegisteredDomains::<T>::insert(domain_lc.clone(), domain_meta);
-            DomainsByOwner::<T>::mutate(
-                &owner, |domains| {
-                    domains.try_push(domain_lc.clone()).expect("qed; too many domains per account")
-                }
-            );
+            DomainsByOwner::<T>::mutate(&owner, |domains| {
+                domains.try_push(domain_lc.clone()).expect("qed; too many domains per account")
+            });
 
             Self::deposit_event(Event::DomainRegistered { who: owner, domain: full_domain });
             Ok(())
         }
 
-        fn do_set_inner_value(
-            full_domain: DomainName<T>,
-            new_value_opt: Option<InnerValueOf<T>>,
+        fn do_set_record(
+            domain: DomainName<T>,
+            key: RecordKey<T>,
+            value_opt: Option<RecordValue<T>>,
             check_ownership: Option<T::AccountId>,
+            reserve_deposit: bool,
         ) -> DispatchResult {
-            let domain_lc = Self::lower_domain_then_bound(&full_domain);
+            let domain_lc = Self::lower_domain_then_bound(&domain);
             let meta = Self::require_domain(domain_lc.clone())?;
+            let owner = meta.owner.clone();
 
             if let Some(should_be_owner) = check_ownership {
                 Self::ensure_allowed_to_update_domain(&meta, &should_be_owner)?;
             }
 
-            ensure!(meta.inner_value != new_value_opt, Error::<T>::InnerValueNotChanged);
-
-            if let Some(old_value) = &meta.inner_value {
-                DomainByInnerValue::<T>::remove(&meta.owner, old_value);
-            }
+            let (old_depositor, old_deposit): (<T as frame_system::pallet::Config>::AccountId, BalanceOf<T>) = if let Some(value_with_deposit_info) =
+                DomainRecords::<T>::get(domain_lc.clone(), key.clone())
+            {
+                (value_with_deposit_info.1, value_with_deposit_info.2)
+            } else {
+                (owner.clone(), 0u32.into())
+            };
 
-            if let Some(new_value) = &new_value_opt {
-                DomainByInnerValue::<T>::insert(&meta.owner, new_value, domain_lc.clone());
-            }
+            let new_deposit: BalanceOf<T> = if reserve_deposit {
+                let new_deposit = Self::calc_record_deposit(key.clone(), value_opt.clone());
 
-            RegisteredDomains::<T>::mutate(&domain_lc, |meta_opt| {
-                if let Some(stored_meta) = meta_opt {
-                    stored_meta.inner_value = new_value_opt;
+                if owner == old_depositor {
+                    Self::try_reserve_deposit(&owner, old_deposit, new_deposit)?;
+                } else {
+                    <T as Config>::Currency::reserve(&owner, new_deposit)?;
+                    let err_amount = <T as Config>::Currency::unreserve(&old_depositor, old_deposit);
+                    debug_assert!(err_amount.is_zero());
```

Comment:
We mostly don't run a chain using unoptimized builds. Maybe when running tests, but it also depends on a developer.
Don't you think it's redundant?

Reply:
We have this check in other parts of the codebase when we do unreserve. It will have zero effect in runtime so that it will cause no issues.

Also in substrate code they use the debug_assert